_posts/2015-07-15-go-http.md

```diff
@@ -0,0 +1,549 @@
+		w.WriteHeader(404)
+		fmt.Fprintln(w, "That's not a supported number!")
+	})
+
+	hl := logger(h)
+	hhs := newHeaderSetter("X-FOO", "BAR")(hl)
+
+	err := http.ListenAndServe(":9999", hhs)
+	log.Fatal(err)
+}
+```
+
+And here's the curl output:
+
+```
+ ~ ▻ curl -i localhost:9999/three
+ HTTP/1.1 200 OK
+ X-Foo: BAR
+ Date: Wed, 01 Jul 2015 00:39:48 GMT
+ Content-Length: 22
+ Content-Type: text/plain; charset=utf-8
+
+ Here's your number: 3
+
+```
+
+`newHeaderSetter` returns a function which accepts and returns an
+`http.Handler`. Calling that returned function with an `http.Handler` then gets
+you an `http.Handler` which will set the header given to `newHeaderSetter`
+before continuing on to the given `http.Handler`.
+
+This may seem like a strange way of organizing this; for this example the
+signature for `newHeaderSetter` could very well have looked like this:
+
+```
+func newHeaderSetter(key, val string, h http.Handler) http.Handler
+```
+
+And that implementation would have worked fine. But it would have been more
+difficult to compose going forward. In the next section I'll show what I mean.
+
+## Composing middleware with alice
+
+[Alice](https://github.com/justinas/alice) is a very simple and convient helper
+for working with middleware using the function signature I've been holding onto.
+Alice is used to create and use chains of middleware. Chains can even be
+appended to each other, giving even further flexibility. Here's our previous
+example with a couple more headers being set, but also using alice to manage the
+added complexity.
+
+```go
+package main
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+
+	"github.com/justinas/alice"
+)
+
+type numberDumper int
+
+func (n numberDumper) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	fmt.Fprintf(w, "Here's your number: %d\n", n)
+}
+
+func logger(h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		log.Printf("%s requested %s", r.RemoteAddr, r.URL)
+		h.ServeHTTP(w, r)
+	})
+}
+
+type headerSetter struct {
+	key, val string
+	handler  http.Handler
+}
+
+func (hs headerSetter) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set(hs.key, hs.val)
+	hs.handler.ServeHTTP(w, r)
+}
+
+func newHeaderSetter(key, val string) func(http.Handler) http.Handler {
+	return func(h http.Handler) http.Handler {
+		return headerSetter{key, val, h}
+	}
+}
+
+func main() {
+	h := http.NewServeMux()
+
+	h.Handle("/one", numberDumper(1))
+	h.Handle("/two", numberDumper(2))
+	h.Handle("/three", numberDumper(3))
+	h.Handle("/four", numberDumper(4))
+
+	fiveHS := newHeaderSetter("X-FIVE", "the best number")
+	h.Handle("/five", fiveHS(numberDumper(5)))
+
+	h.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(404)
+		fmt.Fprintln(w, "That's not a supported number!")
+	})
+
+	chain := alice.New(
+		newHeaderSetter("X-FOO", "BAR"),
+		newHeaderSetter("X-BAZ", "BUZ"),
+		logger,
+	).Then(h)
```

Comment:
true, but that's not something go is going to provide as a language feature. Go happens to have some functional properties, but it's not meant to be a functional language. This is probably the only instance where functional composition has been useful to me in a go project.